extensions/pkg/predicate/preconditions_test.go

```diff
@@ -0,0 +1,174 @@
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+var _ = Describe("Preconditions", func() {
+
+	Describe("IsInGardenNamespacePredicate", func() {
+		var (
+			pred predicate.Predicate
+			obj  *extensionsv1alpha1.Infrastructure
+		)
+
+		BeforeEach(func() {
+			pred = IsInGardenNamespacePredicate
+			obj = &extensionsv1alpha1.Infrastructure{}
+		})
+
+		Describe("#Create, #Update, #Delete, #Generic", func() {
+			tests := func(run func(client.Object) interface{}) {
+				It("should return false because obj is nil", func() {
+					Expect(run(nil)).To(BeFalse())
+				})
+
+				It("should return false because obj is not in garden namespace", func() {
+					obj.SetNamespace("foo")
+					Expect(run(obj)).To(BeFalse())
+				})
+
+				It("should return true because obj is in garden namespace", func() {
+					obj.SetNamespace("garden")
+					Expect(run(obj)).To(BeTrue())
+				})
+			}
+
+			tests(func(obj client.Object) interface{} { return pred.Create(event.CreateEvent{Object: obj}) })
+			tests(func(obj client.Object) interface{} { return pred.Update(event.UpdateEvent{ObjectNew: obj}) })
+			tests(func(obj client.Object) interface{} { return pred.Delete(event.DeleteEvent{Object: obj}) })
+			tests(func(obj client.Object) interface{} { return pred.Generic(event.GenericEvent{Object: obj}) })
+		})
+	})
+
+	Describe("#ShootNotFailedPredicate", func() {
+		var (
+			ctx = context.TODO()
+
+			ctrl  *gomock.Controller
+			cache *mockcache.MockCache
+
+			pred     predicate.Predicate
+			injector = func(predicate interface{}) {
+				Expect(inject.InjectorInto(func(into interface{}) error {
+					Expect(inject.StopChannelInto(ctx.Done(), into)).To(BeTrue())
+					Expect(inject.CacheInto(cache, into)).To(BeTrue())
+					return nil
+				}, predicate)).To(BeTrue())
+			}
+
+			obj       *extensionsv1alpha1.Infrastructure
+			namespace = "shoot--foo--bar"
+		)
+
+		BeforeEach(func() {
+			ctrl = gomock.NewController(GinkgoT())
+			cache = mockcache.NewMockCache(ctrl)
+
+			pred = ShootNotFailedPredicate()
+			injector(pred)
+
+			obj = &extensionsv1alpha1.Infrastructure{ObjectMeta: metav1.ObjectMeta{Namespace: namespace}}
+		})
+
+		AfterEach(func() {
+			ctrl.Finish()
+		})
+
+		Describe("#Create, #Update", func() {
+			tests := func(run func() interface{}) {
+				It("should return true because shoot has no last operation", func() {
+					cache.EXPECT().Get(gomock.Any(), kutil.Key(namespace), gomock.AssignableToTypeOf(&extensionsv1alpha1.Cluster{})).DoAndReturn(func(_ context.Context, _ client.ObjectKey, actual *extensionsv1alpha1.Cluster) error {
```

Comment:
Why not use a fake client? Mocks are not needed here, right?

Reply:
The tests were copied from the old implementation :) But you're right, I'll adapt.